common/src/main/java/com/wynntils/features/user/ChatMentionFeature.java

```diff
@@ -0,0 +1,123 @@
+        pattern = buildPattern();
+    }
+
+    @Override
+    protected void onConfigUpdate(ConfigHolder configHolder) {
+        // rebuild pattern in case it has changed
+        pattern = buildPattern();
+    }
+
+    private Pattern buildPattern() {
+        return Pattern.compile(
+                "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                        + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+                Pattern.CASE_INSENSITIVE);
+    }
+
+    @SubscribeEvent()
+    public void onChat(ChatMessageReceivedEvent e) {
+        Component message = e.getMessage();
+
+        Matcher looseMatcher = pattern.matcher(ComponentUtils.getUnformatted(message));
+
+        if (looseMatcher.find()) {
+            if (markMention) {
+                e.setMessage(rewriteComponentTree(message));
+            }
+            if (dingMention) {
+                McUtils.playSound(SoundEvents.NOTE_BLOCK_PLING.value());
+            }
+        }
+    }
+
+    private Component rewriteComponentTree(Component comp) {
+        // Make a copy of the component without the siblings
+        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
+        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
+        // because wynn still uses legacy coloring for it.
+        String text = curr.getString();
+
+        // if current component has text check it for mentions and rewrite the component if necessary
+        if (text != "") {
+            curr = rewriteMentions(curr, text, comp.getStyle());
+        }
+
+        // process and append siblings
+        for (Component c : comp.getSiblings()) {
+            curr.append(rewriteComponentTree(c));
+        }
+
+        return curr;
+    }
+
+    private MutableComponent rewriteMentions(MutableComponent curr, String text, Style style) {
+        Matcher match = pattern.matcher(text);
+
+        // if we match then rewrite the component if there are no matches the function will just return
+        if (match.find()) {
+            // get the start of the string before any mention and set curr to be it
+            curr = Component.literal(text.substring(0, match.start())).withStyle(style);
+
+            // do the name of the first mention
+            curr.append(Component.literal(rewriteColorCode + match.group()));
+
+            // Store the point at which this match ended
+            int lastEnd = match.end();
+
+            // Handle any additional mentions within the message
+            while (match.find()) {
+                // get the bit in between of matches
+                curr.append(Component.literal(text.substring(lastEnd, match.start())))
+                        .withStyle(style);
+
+                // get the name and recolor it
+                curr.append(Component.literal(rewriteColorCode + match.group()));
```

Comment:
```suggestion
                curr.append(Component.literal(rewriteColorCode + match.group(0)));
```

Reply:
This should actually use .group() instead of .group(0) as we want the specific group which was matched that time and not the first time.

so if we for example have a message
hi name hows it going alias

where name is the playername and alias is an set alias then this code would change the message to
hi name hows it going name

Reply:
```
public String group() {
        return this.group(0);
    }
    ```

Reply:
well in that case I must have misunderstood the documentation